pkg/statsd/aggregator.go

```diff
@@ -161,15 +172,67 @@ func (a *MetricAggregator) Flush(flushInterval time.Duration) {
 			timer.Count = int(round(timer.SampledCount))
 			timer.PerSecond = timer.SampledCount / flushInSeconds
 
+			if contains(timer.Tags, PercentileBucketsMarkerTag) {
+				bucketsSet, err := createBucketsSet(timer.Tags)
+				if err == nil {
+					timer.Buckets = calculatePercentileBuckets(timer, bucketsSet)
+				}
+			}
+
 			a.metricMap.Timers[key][tagsKey] = timer
 		} else {
 			timer.Count = 0
 			timer.SampledCount = 0
-			timer.PerSecond = 0
 		}
 	})
 }
 
+func createBucketsSet(tags gostatsd.Tags) (buckets []int, err error) {
+	if contains(tags, PercentileBucketsPow2Algorithm) {
+		return []int{2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192}, nil
+	} else if contains(tags, PercentileBucketsPow4Algorithm) {
+		return []int{4, 16, 64, 256, 1024, 4096, 16384}, nil
+	} else if contains(tags, PercentileBucketsLinearAlgorithm) {
+		return []int{100, 200, 300, 400, 500, 600, 700, 800, 900, 1000, 1100, 1200, 1300, 1400, 1500, 1600, 1700, 1800, 1900, 2000}, nil
+	} else if contains(tags, PercentileBucketsTailLatencyAlgorithm) {
+		return []int{2, 8, 32, 64, 100, 200, 300, 400, 500, 750, 1000, 1250, 1500, 1750, 2000, 4000, 6000, 8000, 10000, 20000, 30000, 40000, 50000, 60000}, nil
+	} else if contains(tags, PercentileBucketsRoundDoublingAlgorithm) {
+		return []int{10, 20, 50, 100, 200, 500, 1000, 2000, 5000, 10000}, nil
+	} else {
+		return nil, errors.New("no buckets set algorithm specified")
+	}
+}
+
+func calculatePercentileBuckets(timer gostatsd.Timer, bucketsSet []int) map[int]int {
+	var buckets = append(bucketsSet, InfinityBucketSize)
+
+	var bucketToCountMap = *treemap.NewWithIntComparator()
```

Comment:
I'm not a big fan of dragging in the external data structure to hold this.  You know that the timer values will be sorted (or you can make it contractual, either way), which means you can just straight up count the number of values before you reach the next threshold.  This is much more efficient, and removes the ugly `interface{}` based map, and the need to cast everything.